src/bsp/platform/cc3200/xi_bsp_io_fs_cc3200.c

```diff
@@ -0,0 +1,213 @@
+{
+    _i32 result_file_read = 0;
+
+    static _u8 read_buffer[XI_BSP_IO_FS_READ_BUFFER_SIZE];
+
+    if ( resource_handle == firmware_file_handle_last_opened )
+    {
+        result_file_read = sl_extlib_FlcReadFile( resource_handle, offset, read_buffer,
+                                                  XI_BSP_IO_FS_READ_BUFFER_SIZE );
+    }
+    else
+    {
+        result_file_read = sl_FsRead( resource_handle, offset, read_buffer,
+                                      XI_BSP_IO_FS_READ_BUFFER_SIZE );
+    }
+
+    if ( result_file_read < 0 )
+    {
+        return XI_FS_READ_ERROR;
+    }
+    else
+    {
+        *buffer      = read_buffer;
+        *buffer_size = result_file_read;
+
+        return XI_STATE_OK;
+    }
+}
+
+xi_state_t xi_bsp_io_fs_write( const xi_fs_resource_handle_t resource_handle,
+                               const uint8_t* const buffer,
+                               const size_t buffer_size,
+                               const size_t offset,
+                               size_t* const bytes_written )
+{
+    { /* led indicator of file writes */
+#ifdef XI_DEBUG__FOR_FIRMWARE_UPDATE_TESTING_PURPOSES
+        if ( IMG_ACT_USER1 == sBootInfo.ucActiveImg )
+        {
+            GPIO_IF_LedOff( MCU_ORANGE_LED_GPIO );
+            GPIO_IF_LedToggle( MCU_GREEN_LED_GPIO );
+        }
+        else
+        {
+            GPIO_IF_LedOff( MCU_GREEN_LED_GPIO );
+            GPIO_IF_LedToggle( MCU_ORANGE_LED_GPIO );
+        }
+#else
+        GPIO_IF_LedToggle( MCU_GREEN_LED_GPIO );
+#endif
+    }
+
+    /* reasonably NOT supporting simultanenous firmware writes */
+    if ( resource_handle == firmware_file_handle_last_opened )
+    {
+        /* write firmware file */
+        *bytes_written = sl_extlib_FlcWriteFile( resource_handle, offset, ( _u8* )buffer,
+                                                 buffer_size );
+    }
+    else
+    {
+        /* write ordinary file */
+        *bytes_written =
+            sl_FsWrite( resource_handle, offset, ( _u8* )buffer, buffer_size );
+    }
+
+    return ( buffer_size == *bytes_written ) ? XI_STATE_OK : XI_FS_WRITE_ERROR;
+}
+
+xi_state_t xi_bsp_io_fs_close( const xi_fs_resource_handle_t resource_handle )
+{
+    if ( resource_handle == firmware_file_handle_last_opened )
+    {
+        firmware_file_handle_last_opened = 0;
+
+        return ( 0 == sl_extlib_FlcCloseFile( resource_handle, NULL, NULL, 0 ) )
+                   ? XI_STATE_OK
+                   : XI_FS_CLOSE_ERROR;
+    }
+    else
+    {
+        return ( 0 == sl_FsClose( resource_handle, NULL, NULL, 0 ) ) ? XI_STATE_OK
+                                                                     : XI_FS_CLOSE_ERROR;
+    }
+}
+
+xi_state_t xi_bsp_io_fs_remove( const char* const resource_name )
+{
+    return ( 0 == sl_FsDel( ( const _u8* )resource_name, NULL ) ) ? XI_STATE_OK
+                                                                  : XI_FS_ERROR;
```

Comment:
Should some self kill logic reside here? Anyway how to check it? The only way to check it is to call the xi_bsp_fwu_is_this_firmware function. But it tells "yes" only for "firmware.bin" string. And actually the firmware on CC3200 is /sys/mcuimg1.bin or /sys/mcuimg2.bin. And yes, I could test against these strings but maybe the inactive firmware is asked to be removed which might be ok. The above copied function - which checks active FW is private inside the SDK - is just a copy, new SDK could update its body.
So two things: 
1 - I don't know whether it is the responsibility of a FILE BSP to prevent application to be dumb.
2 - Not that easy to deny only active firmware deletion.

So what to do? If you still advice to implement it I can create a heuristic which prevents deletion in 98% of the cases :) Although I think it is an unnecessary thing especially because it is more complex than its value. Especially at this stage of the FWU feature.